mds/management/commands/poll_providers.py

```diff
@@ -0,0 +1,118 @@
+import datetime
+import logging
+import traceback
+import urllib.parse
+
+from django.contrib.gis import geos
+from django.core import management
+from django.db.models.aggregates import Max
+
+import requests
+import pytz
+
+from mds import models
+
+
+NOT_PROVIDED = object()
+
+logger = logging.getLogger(__name__)
+
+
+class Command(management.BaseCommand):
+    help = "Poll providers to fetch their latest data."
+
+    def handle(self, *args, **options):
+        for provider in models.Provider.objects.all():
+            if not provider.base_api_url:
+                self.stdout.write(
+                    "Provider %s has no URL defined, skipping." % provider.name
+                )
+                continue
+
+            self.stdout.write("Polling %s... " % provider.name, ending="")
+            try:
+                self.poll_status_changes(provider)
+            except Exception:  # pylint: disable=broad-except
+                logger.exception("Error in polling provider %s", provider.name)
+                self.stderr.write("Polling failed!")
+                self.stderr.write(traceback.format_exc())
+                # Try the next provider anyway
+                continue
+
+            self.stdout.write("Success!")
+
+    def poll_status_changes(self, provider):
+        next_url = urllib.parse.urljoin(provider.base_api_url, "/status_changes")
+        # Start where we left
+        # (we wouldn't know if the provider recorded late telemetries after the last sync)
+        max_timestamp = models.EventRecord.objects.filter(
+            device__provider=provider
+        ).aggregate(max_timestamp=Max("timestamp"))["max_timestamp"]
+        if max_timestamp:
+            next_url = "%s?%s" % (
+                next_url,
+                urllib.parse.urlencode(
+                    {"start_time": int(max_timestamp.timestamp() * 1000)}
+                ),
+            )
+
+        # Pagination
+        while next_url:
+            data = self.get_data(next_url)
+            for status_change in data["data"]["status_changes"]:
+                device, created = models.Device.objects.get_or_create(
+                    pk=status_change["device_id"],
+                    defaults={
+                        "provider": provider,
+                        "identification_number": status_change["vehicle_id"],
+                        "category": status_change["vehicle_type"],
+                        "propulsion": status_change["propulsion_type"],
+                    },
+                )
```

Comment:
Maybe consider something less linear ? This `get_or_create` does 3-4 queries for each status change...

Doing it in two requests (fetch all devices, create missing) is a challenge because of concurrency considerations. However, doing a fetch first (most devices will already be registered) before trying `get_or_create` only on missing devices will be better.

Alternatively (see below) use PostgreSQL `ON CONFLICT` instruction for upsert.